src/test/java/htsjdk/samtools/SamReaderTest.java

```diff
@@ -169,4 +177,89 @@ public void samRecordFactoryNullHeaderTest() {
         Assert.assertTrue(samRec.getHeader() == null);
     }
 
+    @Test
+    public void testAssertingIteratorUsesLenientOrdering() {
+        //the coordinate comparator's strict sort sorts lower mapping qualities first
+        //so this list is not sorted with respect to that comparator, but it is sorted with respect to the more lenient
+        // file order comparator which only checks the position
+        final List<SAMRecord> looselySorted = Arrays.asList(createRecord(1, 10),
+                                                            createRecord(1, 1),
+                                                            createRecord(2, 1));
+
+        final SAMRecordCoordinateComparator coordinateComparator = new SAMRecordCoordinateComparator();
+
+        //sanity check that this really sorts differently with the file order comparator vs the full ordering coordinate order comparator
+        final List<SAMRecord> sortedWithFileOrderComparator = looselySorted.stream()
+                .sorted(coordinateComparator::fileOrderCompare)
+                .collect(Collectors.toList());
+        Assert.assertEquals(
+                sortedWithFileOrderComparator, looselySorted);
+
+        final List<SAMRecord> sortedWithFullOrderComparator = looselySorted.stream()
+                .sorted(coordinateComparator)
+                .collect(Collectors.toList());
+
+        Assert.assertNotEquals(sortedWithFullOrderComparator, looselySorted);
+
+        final SamReader.AssertingIterator iter = new SamReader.AssertingIterator(
+                wrapInCloseableIterator(looselySorted));
+        iter.assertSorted(SAMFileHeader.SortOrder.coordinate);
+        int count = 0;
+
+        while (iter.hasNext()) {
+            iter.next();
+            count++;
+        }
+        Assert.assertEquals(count, 3);
+    }
+
+    @Test(expectedExceptions = IllegalStateException.class)
+    public void testAssertingIteratorCorrectlyFailsWhenOutOfOrder() {
+        final List<SAMRecord> unsorted = Arrays.asList(createRecord(10, 1), createRecord(1, 1));
+        final SamReader.AssertingIterator iter = new SamReader.AssertingIterator(wrapInCloseableIterator(unsorted));
+        iter.assertSorted(SAMFileHeader.SortOrder.coordinate);
+
+        while (iter.hasNext()) {
+            iter.next();
+        }
+    }
+
+    private static CloseableIterator<SAMRecord> wrapInCloseableIterator(final List<SAMRecord> looselySorted) {
+        return new CloseableIterator<SAMRecord>() {
```

Comment:
A simpler way to do this is to wrap in `PeekableIterator`

Reply:
good call